### PR TITLE
DBZ-1905 Validate database history topic

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.redhat.com/browse/DBZ-1905

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 1]
```diff
@@ -314,6 +327,71 @@ public boolean exists() {
         return exists;
     }
 
+    private void checkTopicSettings(String topicName) {
+        final Thread t = new Thread((Runnable) () -> {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:331]: Could you use `ExecutorService` instead? It should be the preferred way for async task execution.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 2]
```diff
@@ -314,6 +327,71 @@ public boolean exists() {
...
+            try (AdminClient admin = AdminClient.create(this.producerConfig.asProperties())) {
+
+                Set<ConfigResource> resources = Collections.singleton(new ConfigResource(ConfigResource.Type.TOPIC, topicName));
+                final Map<ConfigResource, Config> configs = admin.describeConfigs(resources).all().get(
+                        KAFKA_QUERY_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+                if (configs.size() != 1) {
+                    logger.info("Expected one topic '{}' to match the query but got {}", topicName, configs.values().size());
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:338]: Can you change it to static `LOGGER` while at it?

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 3]
```diff
@@ -56,6 +58,16 @@
 @NotThreadSafe
 public class KafkaDatabaseHistory extends AbstractDatabaseHistory {
 
+    private static final String CLEANUP_POLICY_NAME = "cleanup.policy";
+    private static final String CLEANUP_POLICY_VALUE = "delete";
+    private static final String RETENTION_MS_NAME = "retention.ms";
+    private static final long RETENTION_MS_MAX = Long.MAX_VALUE;
+    private static final long RETENTION_MS_MIN = 157_680_000_000L; // 5 years
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:65]: Let's use `Duration.of(5 , ChronoUnit.YEARS)`, as per the general preference for using the semantical temporal types of Java 8.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java (hunk 1 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:336]: @gunnarmorling I tend to disagree here - the thread is fire and forget. If we want to do the same with the Executor we would need to provide a `ThreadFactory` that does exactly the same - create a thread and set it to daemon and then we should close it to ensure a smmetry in the code. I believe this is a more pragmatic solution given its intent

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java (hunk 3 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Done

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java (hunk 2 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Done

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:336]: The point on thread factory is a good one actually, as it ensures right naming of the thread. See `Threads.newSingleThreadExecutor()`.

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: @gunnarmorling `ThreadFactory` ensure naming only if `setName` is explictly called on the created thread which was done even before. Anyway I've added new commit with rewrite to executors API.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 4]
```diff
@@ -56,6 +63,18 @@
 @NotThreadSafe
 public class KafkaDatabaseHistory extends AbstractDatabaseHistory {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaDatabaseHistory.class);
+
+    private static final String CLEANUP_POLICY_NAME = "cleanup.policy";
+    private static final String CLEANUP_POLICY_VALUE = "delete";
+    private static final String RETENTION_MS_NAME = "retention.ms";
+    private static final long RETENTION_MS_MAX = Long.MAX_VALUE;
+    private static final long RETENTION_MS_MIN = Duration.of(5, ChronoUnit.DAYS).toMillis();
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:72]: Days, not years?

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 5]
```diff
@@ -128,11 +147,21 @@
     private volatile KafkaProducer<String, String> producer;
     private int maxRecoveryAttempts;
     private Duration pollInterval;
+    private final ExecutorService checkTopicSettingsExecutor;
+
+    public KafkaDatabaseHistory() {
+        checkTopicSettingsExecutor = Executors.newSingleThreadExecutor((r) -> {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:153]: Why not using our own `Threads` factory? It takes care of the naming and should be the one facility for spawning new threads.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java (hunk 4 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Damn autocomplete, fixed

**Comment by jpechane** [comment]
Requires incubator PR - https://github.com/debezium/debezium-incubator/pull/155

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 6]
```diff
@@ -108,8 +129,24 @@
             .withDefault(100)
             .withValidation(Field::isInteger);
 
+    public static final Field CONNECTOR_CLASS = Field.create(CONFIGURATION_FIELD_PREFIX_STRING + "connector.class")
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:132]: Why are these two `Field`s now? Wouldn't it suffice to pass in the values when constructing the history object? I.e. when is there a need to configure this differently than for the connector itself?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:132]: @gunnarmorling  The history obejct is constructed using `Configuration.getInstance` so it is not possible to pass params to the constructor. We would need to modify the signature of `configure` method but that would mean that we will push thses values to the implementations that does not need them

**Comment by jpechane** [comment]
@gunnarmorling I've changed the param names to signify they are internal ones and commented on them. Also they ar eno longer required but optional in case anybody is using `KafkaDatabaseHistory` in a specific way.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 7]
```diff
@@ -314,8 +364,76 @@ public boolean exists() {
...
+            }
+            catch (Throwable e) {
+                LOGGER.info("Attempted to validate database history topic but failed", e);
+            }
+        });
+    }
+
     @Override
     public synchronized void stop() {
+        if (checkTopicSettingsExecutor != null) {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java:434]: Ok, last comment on this PR, promised :) Can we move the shutdown right to the topic check (with awaitTermination()), so we don't keep it running for all the time.

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: @gunnarmorling Done, but I don't believe you ;-)

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Let me see... ;-)

**Comment by gunnarmorling** [comment]
Thanks a lot, @jpechane! 🎉

**Pull request merged**